cluster/email.go

```diff
@@ -126,20 +128,90 @@ func SendMail(toName, toEmail, subject, body string) error {
 }
 
 // GetTemplate gets a template from the templates folder and applies the template date
-func GetTemplate(templateFileName string, data interface{}) (*string, error) {
-	// Working Directory
-	wd, err := os.Getwd()
+func GetTemplate(templateName string, data interface{}) (*string, error) {
+	// try to load the template from the db
+	dbTemplate, err := getTemplateFromDB(nil, templateName)
 	if err != nil {
-		return nil, err
-	}
-	t, err := template.ParseFiles(wd + fmt.Sprintf("/cluster/templates/%s.html", templateFileName))
-	if err != nil {
-		return nil, err
-	}
+		// ignore no results error
+		if !strings.Contains(err.Error(), "no rows in result set") {
+			return nil, err
+		}
+	}
+	var t *template.Template
+	if dbTemplate != nil && *dbTemplate != "" {
+		// parse the template from the DB
+		t, err = template.New(templateName).Parse(*dbTemplate)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// if we found no template, load from disk
+		// Working Directory
+		wd, err := os.Getwd()
+		if err != nil {
+			return nil, err
+		}
+		t, err = template.ParseFiles(wd + fmt.Sprintf("/cluster/templates/%s.html", templateName))
+		if err != nil {
+			return nil, err
+		}
+	}
+	// replace the {{.Tokens}} from the template with the provided struct data
 	buf := new(bytes.Buffer)
 	if err = t.Execute(buf, data); err != nil {
 		return nil, err
 	}
 	body := buf.String()
 	return &body, nil
 }
+
+func getTemplateFromDB(ctx *Context, templateName string) (*string, error) {
+	query :=
+		`SELECT 
+				et.template
+			FROM 
+				email_templates et
+			WHERE et.id=$1`
+	// non-transactional query
+	var row *sql.Row
+	// did we got a context? query inside of it
+	if ctx != nil {
+		tx, err := ctx.MainTx()
+		if err != nil {
+			return nil, err
+		}
+		row = tx.QueryRow(query, templateName)
+	} else {
+		// no context? straight to db
+		row = GetInstance().Db.QueryRow(query, templateName)
+	}
+
+	// Save the resulted query on the User struct
+	var template string
+	err := row.Scan(&template)
+	if err != nil {
+		return nil, err
+	}
+	return &template, nil
+}
+
+// SetEmailTemplate upserts a template into the database. If the id is not present the record will be inserted, if it's
+// present it will be updated
+func SetEmailTemplate(ctx *Context, templateName, templateBody string) error {
```

Comment:
templateID is being used not the name
```suggestion
func SetEmailTemplate(ctx *Context, templateID, templateBody string) error {
```